File: phase-2/frontend/services/tasks.ts
// frontend/services/tasks.ts
import { apiClient } from './api';
import type {
  Task as TaskType,
  CreateTaskRequest,
  UpdateTaskRequest,
  TaskResponse as RawTaskResponse,
  TasksResponse as RawTasksResponse
} from '@/types/tasks';

// Define raw response types that match the backend API
interface RawTask {
  id: string;
  title: string;
  description?: string;
  completed: boolean;
  user_id: string;
  created_at: string;
  updated_at: string;
}

interface RawTaskResponse {
  task: RawTask;
}

interface RawTasksResponse {
  tasks: RawTask[];
}

// Helper function to convert from backend format to frontend type
const mapRawTaskToFrontend = (rawTask: RawTask): TaskType => {
  return {
    id: rawTask.id,
    title: rawTask.title,
    description: rawTask.description,
    completed: rawTask.completed,
    user_id: rawTask.user_id,
    created_at: rawTask.created_at,
    updated_at: rawTask.updated_at,
  };
};

class TasksService {
  // Get all tasks for the authenticated user
  async getAll(userId: string): Promise<TaskType[]> {
    try {
      const response = await apiClient.get<RawTasksResponse>(`/api/${userId}/tasks`);
      return response.tasks.map(mapRawTaskToFrontend);
    } catch (error) {
      console.error('Error fetching tasks:', error);
      throw error;
    }
  }

  // Get a specific task by ID
  async getById(userId: string, id: string): Promise<TaskType> {
    try {
      const response = await apiClient.get<RawTaskResponse>(`/api/${userId}/tasks/${id}`);
      return mapRawTaskToFrontend(response.task);
    } catch (error) {
      console.error(`Error fetching task with ID ${id}:`, error);
      throw error;
    }
  }

  // Create a new task
  async create(userId: string, taskData: CreateTaskRequest): Promise<TaskType> {
    try {
      const response = await apiClient.post<RawTaskResponse>(`/api/${userId}/tasks`, taskData);
      return mapRawTaskToFrontend(response.task);
    } catch (error) {
      console.error('Error creating task:', error);
      throw error;
    }
  }

  // Update an existing task
  async update(userId: string, id: string, taskData: UpdateTaskRequest): Promise<TaskType> {
    try {
      const response = await apiClient.put<RawTaskResponse>(`/api/${userId}/tasks/${id}`, taskData);
      return mapRawTaskToFrontend(response.task);
    } catch (error) {
      console.error(`Error updating task with ID ${id}:`, error);
      throw error;
    }
  }

  // Delete a task
  async delete(userId: string, id: string): Promise<void> {
    try {
      await apiClient.delete(`/api/${userId}/tasks/${id}`);
    } catch (error) {
      console.error(`Error deleting task with ID ${id}:`, error);
      throw error;
    }
  }

  // Toggle task completion status
  async toggleCompletion(userId: string, id: string): Promise<TaskType> {
    try {
      const response = await apiClient.patch<RawTaskResponse>(`/api/${userId}/tasks/${id}/complete`, {});
      return mapRawTaskToFrontend(response.task);
    } catch (error) {
      console.error(`Error toggling completion for task with ID ${id}:`, error);
      throw error;
    }
  }
}

// Export a singleton instance of the TasksService
export const tasksService = new TasksService();

// Export individual methods for convenience
export const { getAll, getById, create, update, delete: deleteTask, toggleCompletion } = tasksService;

File: phase-2/src/routes/tasks.py
"""
Task routes for Todo Backend API
"""
from fastapi import APIRouter, Depends, HTTPException, status
from typing import List
from uuid import UUID

from ..models.task import Task
from ..models.schemas import TaskCreate
from ..dependencies import get_db_session
from sqlmodel.ext.asyncio.session import AsyncSession
from sqlmodel import select
from sqlalchemy import desc


router = APIRouter(
    prefix="/api/{user_id}/tasks",
    tags=["tasks"],
    responses={404: {"description": "User or tasks not found"}},
)


@router.get("/", response_model=List[Task])
async def list_tasks(
    user_id: UUID,
    session: AsyncSession = Depends(get_db_session),
    skip: int = 0,
    limit: int = 100,
    sort_by: str = "created_at",
    order: str = "desc"
):
    """
    Get all tasks for a specific user with pagination and sorting
    """
    try:
        from ..services import TaskService

        # Get non-deleted tasks for the user
        tasks = await TaskService.get_non_deleted_tasks_for_user(
            session, user_id, skip, limit, sort_by, order
        )

        return tasks
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error retrieving tasks: {str(e)}"
        )


@router.post("/", response_model=Task)
async def create_task(
    user_id: UUID,
    task_data: TaskCreate,
    session: AsyncSession = Depends(get_db_session)
):
    """
    Create a new task for a specific user
    """
    try:
        # Create new task instance with user_id
        task = Task(
            title=task_data.title,
            description=task_data.description,
            user_id=user_id
        )

        # Add to session and commit
        session.add(task)
        await session.commit()
        await session.refresh(task)

        return task
    except Exception as e:
        await session.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error creating task: {str(e)}"
        )


@router.patch("/{task_id}/complete", response_model=Task)
async def toggle_task_completion(
    user_id: UUID,
    task_id: UUID,
    session: AsyncSession = Depends(get_db_session)
):
    """
    Toggle the completion status of a specific task
    """
    try:
        from ..services import TaskService

        # Validate that task exists and belongs to user
        task = await TaskService.get_task_by_id(session, task_id, user_id)
        if not task:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Task not found or does not belong to user"
            )

        # Toggle completion status
        task.completed = not task.completed
        await session.commit()
        await session.refresh(task)

        return task
    except HTTPException:
        raise
    except Exception as e:
        await session.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error toggling task completion: {str(e)}"
        )


@router.put("/{task_id}", response_model=Task)
async def update_task(
    user_id: UUID,
    task_id: UUID,
    task_data: TaskUpdate,
    session: AsyncSession = Depends(get_db_session)
):
    """
    Update a specific task for a user
    """
    try:
        from ..services import TaskService

        # Attempt to update the task
        updated_task = await TaskService.update_task(session, task_id, user_id, task_data)

        if not updated_task:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Task not found or does not belong to user"
            )

        return updated_task
    except HTTPException:
        raise
    except Exception as e:
        await session.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error updating task: {str(e)}"
        )


@router.delete("/{task_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_task(
    user_id: UUID,
    task_id: UUID,
    session: AsyncSession = Depends(get_db_session)
):
    """
    Delete a specific task for a user
    """
    try:
        from ..services import TaskService

        # Attempt to delete the task
        success = await TaskService.delete_task(session, task_id, user_id)

        if not success:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Task not found or does not belong to user"
            )

        # Return 204 No Content on success
        return
    except HTTPException:
        raise
    except Exception as e:
        await session.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error deleting task: {str(e)}"
        )